src/Build.UnitTests/Evaluation/ProjectSdkImplicitImport_Tests.cs

```diff
@@ -534,6 +542,257 @@ public void IgnoreMissingImportsSdkNotFoundDoesNotThrow()
             }
         }
 
+        internal class SdkPropertiesAreExpandedDataTemplate
+        {
+            public SdkPropertiesAreExpandedDataTemplate(string template, bool expectedMinimumVersionIsNull)
+            {
+                Template = template ?? throw new ArgumentNullException(nameof(template));
+                ExpectedMinimumVersion = expectedMinimumVersionIsNull ? null : string.Empty;
+            }
+
+            public string Template { get; }
+            public string ExpectedMinimumVersion { get; }
+
+            public override string ToString()
+            {
+                return Template switch
+                {
+                    ProjectTemplateSdkAsAttributeWithVersion => nameof(ProjectTemplateSdkAsAttributeWithVersion),
+                    ProjectTemplateSdkAsElementWithVersion => nameof(ProjectTemplateSdkAsElementWithVersion),
+                    ProjectTemplateSdkAsExplicitImportWithVersion => nameof(ProjectTemplateSdkAsExplicitImportWithVersion),
+                    _ => "<unknown>"
+                };
+            }
+        }
+
+        internal class SdkPropertiesAreExpandedCase
+        {
+            public SdkPropertiesAreExpandedCase(SdkReferencePropertyExpansionMode? mode,
+                SdkPropertiesAreExpandedDataTemplate template, bool setName, bool setVersion, bool expectedSuccess)
+            {
+                Mode = mode;
+                Template = template ?? throw new ArgumentNullException(nameof(template));
+                ExpectedSuccess = expectedSuccess;
+                SetNameProperty = setName;
+                SetVersionProperty = setVersion;
+            }
+
+            public SdkReferencePropertyExpansionMode? Mode { get; }
+            public SdkPropertiesAreExpandedDataTemplate Template { get; }
+            public string TemplateName { get; set; } = SdkName;
+            public string TemplateVersion { get; set; } = SdkExpectedVersion;
+            public bool SetNameProperty { get; }
+            public bool SetVersionProperty { get; }
+            public bool ExpectedSuccess { get; }
+
+            public override string ToString()
+            {
+                var result = new StringBuilder(256);
+                if (Mode.HasValue)
+                    result.Append(Mode);
+                else
+                    result.Append($"{nameof(Mode)}: <null>");
+
+                result.Append($", {Template}, {nameof(TemplateName)}: {TemplateName}, {nameof(TemplateVersion)}: {TemplateVersion}");
+
+                if (SetNameProperty)
+                    result.Append(", SetName");
+                if (SetVersionProperty)
+                    result.Append(", SetVersion");
+                if (ExpectedSuccess)
+                    result.Append(", ExpectedSuccess");
+
+                return result.ToString();
+            }
+        }
+
+        [Theory]
```

Comment:
nit:
Does this still need to be a theory?

Reply:
Well, since it is a parametrized test, it has to be a `Theory`, not a `Fact`.

Reply:
I was thinking there's only one input used with the test, so you could inline it.

Reply:
You propose inlining the whole set of subtests from `SdkPropertiesAreExpandedData` into this method? I'm not sure why is that better than having xUnit do the full stats per-subtest.

Reply:
I'm also not sure what you mean by "one input". That's 240 different dynamically-generated inputs.

Reply:
I should have said the data come from one place. That means that converting it to a Fact would mean a single for loop, which isn't too messy. I normally think theories are better if I would need to specify a lot of data in the test to keep it a Fact. In this case, you could remove the `new object[]` part in SdkPropertiesAreExpandedData and make it a simple loop through all the Cases, which isn't ugly.

Reply:
In any case, it isn't bothering me too much, so you can ignore me if you want.

Reply:
1. Will setting the environment variable still work? The value in `EscapeHatches` is set on first access, it can't be changed later, so each change in Mode requires at least reloading the assemblies (if not restarting the test host altogether).
2. Why reimplement the xUnit parametrized test? What's the expected behavior if one subtest fails? Assert will stop the iteration and the log will contain only the fact that something has failed, not the data on which it failed or the amount of successful/failed subtests.

I strongly oppose the suggested inlining.

Reply:
I actually don't know about (1), so if it doesn't, you're probably right. I would have thought `var project = Project.FromXmlReader(xmlReader, projectOptions);` is the point at which the environment variable needs to have changed, since that simulates making a new project.
For (2), I tend to debug into them the same, so it doesn't really matter to me.

You feel far more strongly than I do, so it's fine to leave it as-is. I already approved.